core/src/providers/mistralai.rs

```diff
@@ -0,0 +1,680 @@
+pub enum MistralAIChatMessageRole {
+    Assistant,
+    System,
+    User,
+}
+
+impl From<MistralAIChatMessageRole> for ChatMessageRole {
+    fn from(value: MistralAIChatMessageRole) -> Self {
+        match value {
+            MistralAIChatMessageRole::Assistant => ChatMessageRole::Assistant,
+            MistralAIChatMessageRole::System => ChatMessageRole::System,
+            MistralAIChatMessageRole::User => ChatMessageRole::User,
+        }
+    }
+}
+
+impl ToString for MistralAIChatMessageRole {
+    fn to_string(&self) -> String {
+        match self {
+            MistralAIChatMessageRole::Assistant => String::from("assistant"),
+            MistralAIChatMessageRole::System => String::from("system"),
+            MistralAIChatMessageRole::User => String::from("user"),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ChatChoice {
+    pub finish_reason: Option<String>,
+    pub index: usize,
+    pub message: ChatMessage,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct Usage {
+    pub completion_tokens: Option<u64>,
+    pub prompt_tokens: u64,
+    pub total_tokens: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ChatCompletion {
+    pub id: String,
+    pub object: String,
+    pub created: u64,
+    pub model: String,
+    pub choices: Vec<ChatChoice>,
+    pub usage: Option<Usage>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ChatDelta {
+    pub delta: Value,
+    pub finish_reason: Option<String>,
+    pub index: usize,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ChatChunk {
+    pub choices: Vec<ChatDelta>,
+    pub created: u64,
+    pub id: String,
+    pub model: String,
+    pub object: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct Error {
+    #[serde(alias = "type")]
+    pub _type: Option<String>,
+    pub code: Option<String>,
+    pub message: String,
+    pub object: String,
+    pub param: Option<String>,
+}
+
+impl Error {
+    pub fn message(&self) -> String {
+        format!("MistralAIAPIError: [{:?}] {}", self._type, self.message,)
+    }
+
+    pub fn retryable(&self) -> bool {
+        match self.object.as_str() {
+            "error" => match self._type {
+                Some(_) => self.message.contains("retry"),
+                None => false,
+            },
+            _ => false,
+        }
+    }
+
+    pub fn retryable_streamed(&self) -> bool {
+        self.retryable()
+    }
+}
+
+pub struct MistralAILLM {
+    id: String,
+    api_key: Option<String>,
+}
+
+impl MistralAILLM {
+    pub fn new(id: String) -> Self {
+        MistralAILLM { id, api_key: None }
+    }
+
+    fn chat_uri(&self) -> Result<Uri> {
+        Ok(format!("https://api.mistral.ai/v1/chat/completions",).parse::<Uri>()?)
+    }
+
+    fn tokenizer(&self) -> Arc<Mutex<CoreBPE>> {
+        // TODO(flav): Pending further insights into Mistral tokenizer behavior.
+        return cl100k_base_singleton();
```

Comment:
Can you just check the size of the vocab of Mistral. Maybe we want to use the smaller tokenizer from tiktoken vs 100k to have a better estimate

Reply:
I posted a message on the Slack channel asking about the vocabulary size used in the new Mistral models. I will provide an update here once I receive a response with that information.

Reply:
:+1:

Reply:
Addressed in https://github.com/dust-tt/dust/pull/2920